include/nil/crypto3/detail/pack.hpp

```diff
@@ -1,232 +1,864 @@
-                    std::copy(in, in_e, out);
+
+                /*!
+                 * @brief Generic function that chooses pack function depending on input and output iterator category.
+                 *
+                 * @ingroup pack
+                 *
+                 * @tparam InputIterator
+                 * @tparam OutputIterator
+                 *
+                 * @param first
+                 * @param last
+                 * @param out
+                 *
+                 * @return
+                 */
+                template<typename InputIterator, typename OutputIterator>
+                inline static void pack(InputIterator first, InputIterator last, OutputIterator out) {
+                    typedef typename std::iterator_traits<InputIterator>::iterator_category in_cat;
+                    typedef typename std::iterator_traits<OutputIterator>::iterator_category out_cat;
+                    pack(first, last, in_cat(), out, out_cat());
                 }
+
             };
 
-            template<typename Endianness, int InputBits, int OutputBits>
-            struct real_packer<Endianness, InputBits, OutputBits, true, false> {
+            /*!
+             * @brief This packer deals with the case of equal sizes (i.e. InputValueBits == OutputValueBits)
+             * and different endianness representations (or, speaking informally,
+             * InputEndianness != OutputEndianness). It invokes functions which pack input elements
+             * with ValueBits size represented in InputEndianness endianness into output elements
+             * with the same ValueBits size represented in another OutputEndianness endianness.
+             *
+             * @ingroup pack
+             *
+             * @tparam UnitBits
+             * @tparam InputEndian
+             * @tparam OutputEndian
+             * @tparam ValueBits
+             * @tparam InputType
+             * @tparam OutputType
+             */
+            template<int UnitBits, template<int> class InputEndian, template<int> class OutputEndian,
+                     std::size_t ValueBits, typename InputType, typename OutputType>
+            struct packer<InputEndian<UnitBits>, OutputEndian<UnitBits>, ValueBits, ValueBits, InputType, 
+                          OutputType, false, false, false> {
+
+                typedef InputEndian<UnitBits> InputEndianness;
+                typedef OutputEndian<UnitBits> OutputEndianness;
+
+                typedef unit_reverser<InputEndianness, OutputEndianness, UnitBits> units_reverser;
+                typedef bit_reverser<InputEndianness, OutputEndianness, UnitBits> bits_reverser;
+
+                template<typename InputIterator, typename OutputIterator>
+                inline static void pack_n(InputIterator in, std::size_t in_n, OutputIterator out) {
+
+                    BOOST_STATIC_ASSERT(std::is_same<typename std::iterator_traits<InputIterator>::value_type, 
```

Comment:
In case parent class-selector gets implemented, such a ```STATIC_ASSERT``` will not be required.

Reply:
Currently all static asserts in these places removed.